### PR TITLE
[test] add e2e on device training test

### PR DESCRIPTION
**Pull request by pilkicTT**
Adding MNIST training test which runs everything on device, i.e. fwd/bwd of the model and loss function, as well as the optimizer.

Additionally, modified the `test_lora` test to use optimizer on device.

### CODE REVIEW

File: forge/test/mlir/mnist/training/test_training.py [hunk 1]
```diff
@@ -537,3 +529,72 @@ def test_optimizer_device():
...
+        total_loss = 0
+        for batch_idx, (data, target) in enumerate(train_loader):
+
+            # Create target tensor and leave on CPU.
+            target = nn.functional.one_hot(target, num_classes=10).float()
+
+            # Forward pass (prediction) on device.
+            pred = tt_model(data)[0]
+            golden_pred = framework_model(data)
+            assert compare_with_golden(golden_pred, pred, pcc=0.95)
```
Comment (nvukobratTT) [inline forge/test/mlir/mnist/training/test_training.py:572]: Can you use standard `verify` metho?

**Review by nvukobratTT — APPROVED** [review]
Looks good, one minor comment.

File: forge/test/mlir/mnist/training/test_training.py (hunk 1 above)
Comment (pilkicTT) [reply forge/test/mlir/mnist/training/test_training.py:572]: we need the outputs from the model, to pass into loss function, which the `verify()` function does not provide... we can probably expand the `verify()` method to return the values, but I would leave that for later...

Comment (nvukobratTT) [reply forge/test/mlir/mnist/training/test_training.py:572]: https://github.com/tenstorrent/tt-forge-fe/issues/1331

**Review by pglusacTT — APPROVED** [review]
Great, just one question

File: forge/test/mlir/mnist/training/test_training.py [hunk 2]
```diff
@@ -451,16 +441,20 @@ def test_lora():
...
             tt_loss.backward()
 
-            # Adjust weights (on CPU)
-            framework_optimizer.step()
+            # Adjust weights on the device.
+            # NOTE: after executing the step, this will also zero the gradients.
+            tt_optimizer.step()
 
         print(f"epoch: {epoch_idx} loss: {total_loss}")
+        assert prev_total_loss - total_loss > 1.0, "Loss should go down"
```
[the remark this answers is not in the document]
Comment (pglusacTT) [reply forge/test/mlir/mnist/training/test_training.py:456]: If our main concern is floating point issues, I suggest doing something like `assert prev_total_loss > total_loss + 1e-5`, what do you think of that? I think we should be safe in most cases by doing this

[the remark this answers is not in the document]
Comment (pilkicTT) [reply forge/test/mlir/mnist/training/test_training.py]: Done!
